### PR TITLE
Async redis refactor

### PR DESCRIPTION
**Pull request by tylerhutcherson**
- Factors out redis-specific methods to use async client.
- Wraps synchronous methods with asyncio event loop calls.
- Adds more test coverage.

### CODE REVIEW

File: tests/integration_tests/vectorstores/test_redis.py [hunk 1]
```diff
@@ -19,8 +23,42 @@ def test_redis_new_vector() -> None:
     """Test adding a new document"""
     texts = ["foo", "bar", "baz"]
     docsearch = Redis.from_texts(
-        texts, FakeEmbeddings(), redis_url="redis://localhost:6379"
+        texts, FakeEmbeddings(), redis_url=TEST_REDIS_URL
+    )
+    docsearch.add_texts(["foo"])
+    output = docsearch.similarity_search("foo", k=2)
+    assert output == [Document(page_content="foo"), Document(page_content="foo")]
+
```
Comment (agola11) [inline tests/integration_tests/vectorstores/test_redis.py:31]: If there's any async functionality not wrapped in a sync method you can test directly using `pytest.mark.asyncio`

**Review by agola11 — COMMENTED** [review]
thanks! overall looks good, I think some stuff can be simplified by the just calling the entire async coroutine directly from a sync method (i.e. calling `self.loop.run_until_complete(aadd_texts(...))` from `add_texts)

File: tests/integration_tests/vectorstores/test_redis.py (hunk 1 above)
Comment (tylerhutcherson) [reply tests/integration_tests/vectorstores/test_redis.py:31]: I'll add some more tests :)

**Comment by tylerhutcherson** [comment]
@agola11 I pushed up a commit to address your feedback and add new async tests. However, now am getting more event loop issues. This clearly needs to be further tested before we release. If you get a chance, some help here from you would be amazing. I'll be able to come back to this later this weekend.

File: langchain/vectorstores/redis.py [hunk 2]
```diff
@@ -1,128 +1,308 @@
...
+        # Write data to redis in a pipeline
+        pipeline = await self.client.pipeline(transaction=False)
         for i, text in enumerate(texts):
             # Use provided key otherwise use default key
-            key = keys[i] if keys else _redis_key(prefix)
+            key = keys[i] if keys else self._redis_key(prefix)
             metadata = metadatas[i] if metadatas else {}
-            pipeline.hset(
+            embedding = embeddings[i] if embeddings else self.embedding_function(text)
+            await pipeline.hset(
```
Comment (jbkoh) [inline langchain/vectorstores/redis.py:251]: I might be wrong, but I believe in async mode, `hset` doesn't have to be awaited everytime but only at `exceute`
https://redis-py.readthedocs.io/en/stable/connections.html#clusterpipeline-async

**Comment by baskaryan** [comment]
@tylerhutcherson is this still relevant or can be closed?

**Comment by leo-gan** [comment]
@tylerhutcherson Hi , could you, please, resolve the merging issues and address the last comments (if needed)? After that ping me and I push this PR for the review. Thanks!

If this PR is not needed anymore, could you, please, let me know?

**Pull request closed without merging**